src/domain/entities/Questionnaire/QuestionnaireQuestion.ts

```diff
@@ -152,100 +152,102 @@ export class QuestionnaireQuestion {
-            })
-        )
-            .compact()
-            .uniq()
-            .value();
-
         const parsedAndUpdatedQuestions = updatedQuestions.map(question => {
-            //Get the rules that are applicable for the current question
-            //this is done to take care of any "side-effects" of an updated question.
-            const rulesApplicableForCurrentQuestion =
-                question.id !== updatedQuestion.id
-                    ? getApplicableRules(question, questionnaire.rules, updatedQuestions)
-                    : rules;
-
-            //If the question is part of any of the rule actions, update the section
             const parsedAndUpdatedQuestion =
-                allQuestionsRequiringUpdate.includes(question.id) ||
                 question.id === updatedQuestion.id
-                    ? this.updateQuestion(question, rulesApplicableForCurrentQuestion)
+                    ? this.updateQuestion(question, rules)
                     : question;
 
             return parsedAndUpdatedQuestion;
         });
-        //If any of the updated question has been changed to hidden, then reset its value
-        //When it is shown again, the user can enter a "fresh" value
-        const hiddenQuestions = parsedAndUpdatedQuestions.filter(
-            q =>
-                q.isVisible === false &&
-                updatedQuestions.find(uq => uq.id === q.id)?.isVisible === true
-        );
-
-        if (hiddenQuestions.length === 0)
-            return _(parsedAndUpdatedQuestions)
-                .sortBy(question => question.sortOrder)
-                .value();
-
-        const resetQuestions = hiddenQuestions.reduce((acc, hiddenQuestion) => {
-            return this.updateQuestions(
-                acc,
-                { ...hiddenQuestion, value: undefined },
-                rules,
-                questionnaire
-            );
-        }, parsedAndUpdatedQuestions);
-
-        return _(resetQuestions)
+        const sortedUpdatedQuestions = _(parsedAndUpdatedQuestions)
             .sortBy(question => question.sortOrder)
             .value();
+        processedQuestions.push(updatedQuestion);
+
+        //2. Get all questions that need to be updated as a side effect of the current question update
+        const allQuestionIdsRequiringUpdate = _(
+            questionnaire.rules.flatMap(rule => {
+                if (
+                    rule.dataElementIds.includes(updatedQuestion.id) ||
+                    rule.teAttributeIds.includes(updatedQuestion.id)
+                ) {
+                    const actionUpdates = rule.actions.flatMap(
+                        action => action?.dataElement?.id || action.trackedEntityAttribute?.id
+                    );
+
+                    return actionUpdates;
+                }
+            })
+        )
+            .compact()
+            .filter(id => processedQuestions.find(q => q.id === id) === undefined)
+            .uniq()
+            .value();
```

Comment:
This method is very large. For example, here, note that you added a comment of a block of code, so it seems a method/function called `getQuestionsToUpdateAsSideEffect(...)` would make sense. 

Sometimes, we are uneasy about having so many arguments moved between functions when we refactor. In this case, group them in a parent object. (general advice)